### ISSUE TITLE
Resuming training gives different model result / weights

### ISSUE THREAD

**Issue by hiandersson** [Lightning-AI/pytorch-lightning#17543]
### Expected behavior

Saving a checkpoint and resuming from that checkpoint in Pytorch Lightning (2.0.2) should give the same model in the end (weights, states, loss).

### Actual behavior

Saving a checkpoint and resuming from that checkpoint in Pytorch Lightning (2.0.2) gives different training result in the end (checked via MD5, loss etc). This brakes continual learning.

### What version are you seeing the problem on?

v2_0

### How to reproduce the bug

```python
import os
import warnings

warnings.filterwarnings("ignore")  # avoid printing out absolute paths

import copy
from pathlib import Path
import warnings
import tempfile
import json, hashlib

import numpy as np
import pandas as pd
import torch
import random
import lightning.pytorch  as pl

from pytorch_forecasting import TimeSeriesDataSet, NHiTS

import logging
logging.getLogger("lightning.pytorch.accelerators.cuda").setLevel(logging.WARNING)

##################### Eval

class NHitsRunner():

	def __init__(self):

		self.batch_size = 256 
		self.reset_seed()
		self.create_datasets()
		self.create_model()
		self.path = None

	def create_model(self):

		# Dataset

		self.nhits_model = NHiTS.from_dataset(
			self.test_datasets[0],
			learning_rate=0.1,
			hidden_size=16,
			optimizer='adam'
		)

		self.trainer = pl.Trainer(
			devices=1, accelerator="gpu",
			max_epochs=1,
			gradient_clip_val=0.1,
			enable_model_summary=False,
			enable_progress_bar=True,
			enable_checkpointing=False,
			logger=False,
		)

	def create_dataset(self, frame):

		self.max_prediction_length = 10
		self.max_encoder_length = 24
		training_cutoff = frame["time_idx"].max() - self.max_prediction_length

		return TimeSeriesDataSet(
			frame[lambda x: x.time_idx <= training_cutoff],
			time_idx="time_idx",
			target="price",
			group_ids=["group_id"],
			max_encoder_length=self.max_encoder_length,
			max_prediction_length=self.max_prediction_length,
			static_categoricals=[],
			time_varying_known_reals=[],
			time_varying_unknown_categoricals=[],
			time_varying_unknown_reals=[
				"price",
				"volume",
			],
			target_normalizer=None,
		)

	def train_epoch(self, set_id):

		train_dataloader = self.test_datasets[set_id].to_dataloader(train=True, batch_size=self.batch_size, num_workers=0)

		if self.path != None:

			print('Consumed trainer checkpoint')

			self.trainer.fit(
				self.nhits_model,
				ckpt_path=self.path,
				train_dataloaders=train_dataloader,
			)

			os.remove(self.path)

			self.path = None

		else:
			
			self.trainer.fit(
				self.nhits_model,
				train_dataloaders=train_dataloader,
			)

		self.trainer.fit_loop.max_epochs += int(1)

	def reset_seed(self):

		pl.seed_everything(0)
		torch.set_float32_matmul_precision('medium')

	def create_frame(self):

		# Create sine wave
		dataframe_size = 128 * 64
		test_dataframe = pd.DataFrame({'time_idx': np.arange(0, dataframe_size, 1, dtype=int)})
		test_dataframe = test_dataframe.assign(price=lambda x: np.sin(x.time_idx/3)*100)
		test_dataframe = test_dataframe.assign(volume=lambda x: np.sin(x.time_idx/2)*200)
		test_dataframe['group_id'] = 'group_a'
		return test_dataframe

	def create_datasets(self):

		self.test_frames = []
		self.test_datasets = []

		for i in range(3):

			new_frame = self.create_frame()

			self.test_frames.append(new_frame)

			self.test_datasets.append(self.create_dataset(new_frame))

	def convert_dict(self, d, flat_dict):
		for k, v in d.items():
			if isinstance(v, dict):
				self.convert_dict(v, flat_dict)
			else:
				if isinstance(v, torch.Tensor):
					v = v.cpu().detach().numpy().tolist()
				flat_dict[k] = v

	def dict_md5(self):

		_dict = self.nhits_model.state_dict()
		
		flat_dict = {}
		self.convert_dict(_dict, flat_dict)

		data_md5 = hashlib.md5(json.dumps(flat_dict, sort_keys=True).encode('utf-8')).hexdigest()

		return data_md5

	def test_save_load_trainer_checkpoint(self):

		print("Saving trainer checkpoint")

		# Save trainer file

		model_filename = "test_checkpoint_forecast.ckpt"
		tempdir = tempfile.gettempdir()
		file_name = tempdir+"/"+model_filename
		self.trainer.save_checkpoint(file_name)

		self.path = file_name

#### Test running WITHOUT saving and loading 

runner_1 = NHitsRunner()

for i in range(3):

	runner_1.train_epoch(i)

run1_md5 = runner_1.dict_md5()

#### Test running WITHOUt saving and loading 

runner_2 = NHitsRunner()

for i in range(3):

	if i == 2:
		runner_2.test_save_load_trainer_checkpoint()

	runner_2.train_epoch(i)

run2_md5 = runner_2.dict_md5()

#### Compare

if run1_md5 != run2_md5:

	print('Models md5 differs {} != {}'.format(run1_md5, run2_md5))

else:

	print('Models md5 equals {} == {}'.format(run1_md5, run2_md5))
```

### Error messages and logs

```
Global seed set to 0
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.52it/s, train_loss_step=0.737, train_loss_epoch=1.700]`Trainer.fit` stopped: `max_epochs=1` reached.
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.51it/s, train_loss_step=0.737, train_loss_epoch=1.700]
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.75it/s, train_loss_step=0.648, train_loss_epoch=1.150]`Trainer.fit` stopped: `max_epochs=2` reached.
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.74it/s, train_loss_step=0.648, train_loss_epoch=1.150]
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.43it/s, train_loss_step=0.598, train_loss_epoch=1.440]`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.41it/s, train_loss_step=0.598, train_loss_epoch=1.440]
Global seed set to 0
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.57it/s, train_loss_step=0.737, train_loss_epoch=1.700]`Trainer.fit` stopped: `max_epochs=1` reached.
Epoch 0: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 17.55it/s, train_loss_step=0.737, train_loss_epoch=1.700]
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.70it/s, train_loss_step=0.648, train_loss_epoch=1.150]`Trainer.fit` stopped: `max_epochs=2` reached.
Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.69it/s, train_loss_step=0.648, train_loss_epoch=1.150]
Saving trainer checkpoint
Consumed trainer checkpoint
Restoring states from the checkpoint path at /tmp/test_checkpoint_forecast.ckpt
Restored all states from the checkpoint at /tmp/test_checkpoint_forecast.ckpt
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.91it/s, train_loss_step=0.342, train_loss_epoch=0.470]`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████████████| 31/31 [00:01<00:00, 18.90it/s, train_loss_step=0.342, train_loss_epoch=0.470]
Models md5 differs 0d6210c3c296c7acfccac4a4b27e1a16 != 25eee25731cd9d21e9991afa47a90b82

```

### Environment

<details>
  <summary>Current environment</summary>

```
- PyTorch-Forecasting version: 1.0.0 (based on LightningModule)
- PyTorch version: 2.0.0
- PyTorch lightning: 2.0.2 
- Python version: 3.8.10
- Operating System: Ubuntu 20.04
- CUDA 12.1
- Running environment: Local
```

</details>

### More info

In the example code I create 3 data frames and train on each, one epoch at the time. This will render a md5 sum. Rerunning the code will render the same md5 sum.

The second time I run the exact same code, before the last iteration, I save a checkpoint to disk and then resume training from that checkpoint. This will render a different md5 sum. This should not be the case.

**Comment by hiandersson** [comment]
Trying a simple example from Pytorch Lightning "Hello world" example with CIFAR 10 seems to work, the MD5 sum is same when resuming training.
The strange thing is that first I do Trainer.fit with the checkpoint, then `self.trainer.fit_loop.max_epochs += int(1)`, then I train again with Trainer.fit again to get another epoch. Is it suppose to be this way or do I miss something?

```
import os

import lightning as L
import pandas as pd
import seaborn as sn
import torch
from IPython.display import display
from lightning.pytorch.loggers import CSVLogger
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split
from torchmetrics import Accuracy
from torchvision import transforms
from torchvision.datasets import MNIST
import hashlib, json
import logging
logging.getLogger("pytorch_lightning").setLevel(logging.ERROR)
logging.getLogger("lightning.pytorch.utilities.rank_zero").setLevel(logging.WARNING)
logging.getLogger("lightning.pytorch.accelerators.cuda").setLevel(logging.WARNING)

PATH_DATASETS = os.environ.get("PATH_DATASETS", ".")
BATCH_SIZE = 512

class MNISTModel(L.LightningModule):
	def __init__(self):
		super().__init__()
		self.l1 = torch.nn.Linear(28 * 28, 10)

	def forward(self, x):
		return torch.relu(self.l1(x.view(x.size(0), -1)))

	def training_step(self, batch, batch_nb):
		x, y = batch
		loss = F.cross_entropy(self(x), y)
		return loss

	def configure_optimizers(self):
		return torch.optim.Adam(self.parameters(), lr=0.02)
	

class MNISTTest():

	def convert_dict(self, d, flat_dict):
		for k, v in d.items():
			if isinstance(v, dict):
				self.convert_dict(v, flat_dict)
			else:
				if isinstance(v, torch.Tensor):
					v = v.cpu().detach().numpy().tolist()
				flat_dict[k] = v

	def dict_md5(self, model):

		_dict = model.state_dict()
		
		flat_dict = {}
		self.convert_dict(_dict, flat_dict)

		data_md5 = hashlib.md5(json.dumps(flat_dict, sort_keys=True).encode('utf-8')).hexdigest()

		return data_md5
	
	def train(self, resume_checkpoint):

		# Seed    
		L.seed_everything(0)

		# Init our model
		self.mnist_model = MNISTModel()

		# Init DataLoader from MNIST Dataset
		train_ds = MNIST(PATH_DATASETS, train=True, download=True, transform=transforms.ToTensor())
		self.train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE)

		# Initialize a trainer
		self.trainer = L.Trainer(
			accelerator="auto",
			devices=1,
			max_epochs=1,
			enable_model_summary=False,
			enable_progress_bar=True,
			enable_checkpointing=False,
			logger=False,
		)

		# Train the model :zap:
		if resume_checkpoint == None:
		
			self.trainer.fit(self.mnist_model, self.train_loader)

		else:

			print('Resuming from checkpoint {}'.format(resume_checkpoint))

			# self.mnist_model = MNISTModel.load_from_checkpoint(resume_checkpoint)

			self.trainer.fit(
				self.mnist_model, 
				self.train_loader,
				ckpt_path=resume_checkpoint)

	def save(self, file_name):

		self.trainer.save_checkpoint(file_name)

	def get_md5(self):

		return self.dict_md5(self.mnist_model)

	def train_again(self):

		self.trainer.fit_loop.max_epochs += int(1)
		self.trainer.fit(self.mnist_model, self.train_loader)


# Test train again
test_first_run = MNISTTest()
test_first_run.train(None)
test_first_run.save('checkpoint.ckpt')
print('First run Epoch 1: {}'.format(test_first_run.get_md5()))
test_first_run.train_again()
print('First run Epoch 2: {}'.format(test_first_run.get_md5()))

# Test resume from Epoch 1, training one more epoch
test_resume = MNISTTest()
test_resume.train('checkpoint.ckpt')
test_resume.train_again()
print('Train resume Epoch 2: {}'.format(test_resume.get_md5()))

# Clean
#os.remove('checkpoint.ckpt')
```

**Comment by NikolasWolke** [comment]
I did observe the same problem, that a training with resume does not lead to the same result as a training without interruption - even if `deterministic=True`.

But in addition, I see that the global step is higher for the resumed training (48 steps for the reference model without resume but 49 steps for the model with resume).

```python
# ---
# jupyter:
#   jupytext:
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.14.7
#   kernelspec:
#     display_name: Python 3 (ipykernel)
#     language: python
#     name: python3
# ---

# %%
from typing import Any
from copy import deepcopy

import torch
from torch import nn
from torch.nn import functional as F
from torch.optim import AdamW, Optimizer
from torch.optim.lr_scheduler import OneCycleLR
from torch.utils.data import Dataset, DataLoader
from pytorch_lightning import Trainer, LightningModule, seed_everything
from pytorch_lightning.callbacks import Callback, ModelCheckpoint


# %%
class Linear(LightningModule):
    def __init__(self, inputs: int, learning_rate: float):
        super().__init__()
        self.linear = nn.Linear(inputs, 1)
        self.learning_rate = learning_rate

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.linear(x)
    
    def training_step(self, batch: dict[str, torch.Tensor]) -> torch.Tensor:
        x = batch["X"]    # shape (batch_size, inputs)
        y = batch["y"]    # shape (batch_size,)
        y_hat = self.forward(x)
        
        loss = F.mse_loss(y, y_hat)
        return loss
    
    def configure_optimizers(self) -> tuple[list[Optimizer], list[dict[str, Any]]]:
        optimizer = AdamW(self.parameters(), lr=self.learning_rate)
        lr_scheduler = OneCycleLR(
            optimizer=optimizer,
            max_lr=3*self.learning_rate,
            total_steps=self.trainer.estimated_stepping_batches + 1, # TODO: note that since resume does one step more, we need to increase this
        )
        return [optimizer], [{"scheduler": lr_scheduler, "frequency": 1, "interval": "step"}]


# %%
class InterruptTraining(Callback):
    def __init__(self, stop_at_step: int):
        self.stop_at_step = stop_at_step
        self.has_stopped = False
    
    def on_train_epoch_start(self, trainer: Trainer, pl_module: LightningModule) -> None:
        if trainer.global_step == self.stop_at_step and not self.has_stopped:
            self.has_stopped = True
            raise RuntimeError(f"Simulated training termination.")


# %%
class RandomDataset(Dataset):
    def __init__(self, inputs: int, length: int):
        self.len = length
        self.x = torch.randn(length, inputs)
        self.y = 2 * self.x.sum(dim=1) + 3

    def __getitem__(self, index: int) -> dict[str, torch.Tensor]:
        return {"X": self.x[index], "y": torch.zeros(1)}

    def __len__(self) -> int:
        return self.len


# %%
dataset_size = 2**12
batch_size = 256

steps_per_epoch = int(dataset_size/batch_size)
steps_per_epoch

# %%
seed_everything()

dataset = RandomDataset(inputs=10, length=dataset_size)
dataloader = DataLoader(dataset, batch_size=batch_size)

# %%
model = Linear(inputs=10, learning_rate=1e-3)
model_reference = deepcopy(model)

trainer = Trainer(
    max_epochs=3,
    callbacks=[
        InterruptTraining(stop_at_step=steps_per_epoch*2), # simulate training crash after epoch 2
        ModelCheckpoint(every_n_train_steps=1, save_last=True)
    ],
    deterministic=True,
)
try:
    trainer.fit(model, train_dataloaders=dataloader)
except RuntimeError:
    print("Simulated training error --> need to resume")
    trainer.fit(model, train_dataloaders=dataloader, ckpt_path="./lightning_logs/version_0/checkpoints/last.ckpt")

# %%
model.global_step

# %%
model.state_dict()

# %%
trainer = Trainer(
    max_epochs=3,
    callbacks=[
        ModelCheckpoint(every_n_train_steps=1, save_last=True)
    ],
    deterministic=True,
)
trainer.fit(model_reference, train_dataloaders=dataloader)

# %%
model_reference.global_step

# %%
model_reference.state_dict()

# %%
[torch.allclose(p, p_reference) for p, p_reference in zip(model.parameters(), model_reference.parameters())]

# %%

```

**Comment by awaelchli** [comment]
Hey, just wanted to comment here that full fault-tolerance would be nice, but can't be done in most cases. Lightning does and cannot provide any guaratnees of full fault-tolerance. Every time you resume from a checkpoint you start from a different random seed than what was the state before ending the previous training. This means you get different shuffling as if you had not interrupted training. But this is not an issue in practice in most cases because models are not sensitive to initialization/seed/shuffling order. If they were, it would be a different concern anyway. 

So in summary, your check for the md5sum of the model state shouldn't be used to determine whether a model is trained well after resuming from a checkpoint. I suggest that you instead compute metrics on a test or training set to determine the performance of your model instead.
